File: go/deliberation/pkg/deliberation/specialists/polymarket/polymarket.go
// Package polymarket — Go port of @accretive-substrate/deliberation/specialists/polymarket.
//
// Filters venue=polymarket. Reads odds_shift_24h_pp + volume_24h_usd +
// time_to_resolution_days + odds_volatility_pct.
package polymarket

import (
	"fmt"
	"math"
	"strings"

	"github.com/ramene/accretive-substrate/go/deliberation/pkg/deliberation"
)

const Name = "polymarket"

type polyMarket struct {
	OddsShift24hPP       float64 `json:"odds_shift_24h_pp"`
	Volume24hUSD         float64 `json:"volume_24h_usd"`
	TimeToResolutionDays float64 `json:"time_to_resolution_days"`
	OddsVolatilityPct    float64 `json:"odds_volatility_pct"`
}

func Argue(pkg deliberation.EvidencePackage) deliberation.Voice {
	venue := strings.ToLower(pkg.Trigger.Venue)
	if venue != "polymarket" {
		return deliberation.AbstainBecause(Name, fmt.Sprintf("venue=%s out of scope", venue))
	}

	var market polyMarket
	_ = deliberation.DecodeMarketStateFor(pkg.MarketState, pkg.Trigger.Symbol, &market)

	citations := []deliberation.Citation{{
		Type:   "signal",
		Source: "polymarket",
	}}

	if market.OddsShift24hPP == 0 {
		return deliberation.AbstainBecause(Name, "no odds shift data")
	}

	// Near resolution + high volatility — risk off.
	if market.TimeToResolutionDays > 0 && market.TimeToResolutionDays < 1 && market.OddsVolatilityPct > 20 {
		conf := math.Min(0.82, 0.6+market.OddsVolatilityPct/100)
		return deliberation.VoteFor(Name, deliberation.VerdictSell, conf,
			fmt.Sprintf("resolution in %.1fh + odds volatility %.0f%% — risk off",
				market.TimeToResolutionDays*24, market.OddsVolatilityPct),
			citations)
	}

	// Strong momentum + liquidity → buy.
	if market.OddsShift24hPP > 5 && market.Volume24hUSD > 100_000 {
		conf := math.Min(0.82, 0.55+market.OddsShift24hPP/30)
		return deliberation.VoteFor(Name, deliberation.VerdictBuy, conf,
			fmt.Sprintf("odds shifted +%.1fpp on $%dk 24h volume",
				market.OddsShift24hPP, int(market.Volume24hUSD/1000)),
			citations)
	}

	// Event drifting away → sell.
	if market.OddsShift24hPP < -5 {
		conf := math.Min(0.82, 0.55+math.Abs(market.OddsShift24hPP)/30)
		return deliberation.VoteFor(Name, deliberation.VerdictSell, conf,
			fmt.Sprintf("odds shifted %.1fpp — event drifting away", market.OddsShift24hPP),
			citations)
	}

	volLabel := "n/a"
	if market.Volume24hUSD > 0 {
		volLabel = fmt.Sprintf("$%dk", int(market.Volume24hUSD/1000))
	}
	return deliberation.VoteFor(Name, deliberation.VerdictHold, 0.45,
		fmt.Sprintf("shift %.1fpp volume %s — no decisive move", market.OddsShift24hPP, volLabel),
		citations)
}

// Specialist — exported safe-wrapped argue function.
var Specialist = deliberation.MakeSafe(Name, Argue)

File: node/deliberation/src/specialists/polymarket.mjs
/**
 * @accretive-substrate/deliberation/specialists/polymarket
 *
 * Polymarket specialist. Filters venue=polymarket; reads event odds shift
 * + market depth.
 *
 * Verdict rules:
 *   odds_shift_24h > +5pp + volume > $100k        → buy   (momentum + liquidity)
 *   odds_shift_24h < -5pp                          → sell  (event drifting away)
 *   time_to_resolution < 1d AND odds_volatility>20 → sell  (taking risk off near resolution)
 *   else                                            → hold
 *
 * Domain: 'polymarket'.
 */

import { makeSpecialist, voteFor, abstainBecause } from './base.mjs';

export const POLYMARKET_SPECIALIST_NAME = 'polymarket';

async function argue(pkg) {
  const venue = String(pkg?.trigger?.venue || '').toLowerCase();
  if (venue !== 'polymarket') {
    return abstainBecause(POLYMARKET_SPECIALIST_NAME, `venue=${venue} out of scope`);
  }

  const market = pkg?.market_state || {};
  const oddsShift = Number(market.odds_shift_24h_pp);
  const volume = Number(market.volume_24h_usd);
  const timeTo = Number(market.time_to_resolution_days);
  const oddsVol = Number(market.odds_volatility_pct);

  const citations = [{ type: 'signal', source: 'polymarket', market }];

  if (!Number.isFinite(oddsShift)) {
    return abstainBecause(POLYMARKET_SPECIALIST_NAME, 'no odds shift data');
  }

  // Near resolution + high volatility = take risk off.
  if (Number.isFinite(timeTo) && timeTo < 1 && Number.isFinite(oddsVol) && oddsVol > 20) {
    return voteFor(
      POLYMARKET_SPECIALIST_NAME, 'sell', Math.min(0.82, 0.6 + oddsVol / 100),
      `resolution in ${(timeTo*24).toFixed(1)}h + odds volatility ${oddsVol.toFixed(0)}% — risk off`,
      citations,
    );
  }

  // Strong momentum + liquidity = entry.
  if (oddsShift > 5 && Number.isFinite(volume) && volume > 100_000) {
    return voteFor(
      POLYMARKET_SPECIALIST_NAME, 'buy', Math.min(0.82, 0.55 + oddsShift / 30),
      `odds shifted +${oddsShift.toFixed(1)}pp on $${Math.round(volume/1000)}k 24h volume`,
      citations,
    );
  }

  // Event drifting away.
  if (oddsShift < -5) {
    return voteFor(
      POLYMARKET_SPECIALIST_NAME, 'sell', Math.min(0.82, 0.55 + Math.abs(oddsShift) / 30),
      `odds shifted ${oddsShift.toFixed(1)}pp — event drifting away`,
      citations,
    );
  }

  return voteFor(
    POLYMARKET_SPECIALIST_NAME, 'hold', 0.45,
    `shift ${oddsShift.toFixed(1)}pp volume ${Number.isFinite(volume) ? '$'+Math.round(volume/1000)+'k' : 'n/a'} — no decisive move`,
    citations,
  );
}

export const polymarketSpecialist = makeSpecialist(
  POLYMARKET_SPECIALIST_NAME,
  ['polymarket'],
  argue,
);
